sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -396,45 +405,64 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
-            // the default gateway producer can be used to request automatic routing from the Event Hubs service gateway.
+            options ??= DefaultSendOptions;
 
-            TransportProducer activeProducer;
+            AssertSinglePartitionReference(options.PartitionId, options.PartitionKey);
 
-            if (string.IsNullOrEmpty(options.PartitionId))
-            {
-                activeProducer = EventHubProducer;
-            }
-            else
+            while (!isMessageSent)
             {
-                // This assertion is intended as an additional check, not as a guarantee.  There still exists a benign
-                // race condition where a transport producer may be created after the client has been closed; in this case
-                // the transport producer will be force-closed with the associated connection or, worst case, will close once
-                // its idle timeout period elapses.
+                // Determine the transport producer to delegate the send operation to.  Because sending to a specific
+                // partition requires a dedicated client, use (or create) that client if a partition was specified.  Otherwise
+                // the default gateway producer can be used to request automatic routing from the Event Hubs service gateway.
 
-                Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));
-                activeProducer = PartitionProducers.GetOrAdd(options.PartitionId, id => Connection.CreateTransportProducer(id, RetryPolicy));
-            }
+                TransportProducer activeProducer;
 
-            using DiagnosticScope scope = CreateDiagnosticScope();
+                if (string.IsNullOrEmpty(options.PartitionId))
+                {
+                    activeProducer = EventHubProducer;
+                }
+                else
+                {
+                    // This assertion is intended as an additional check, not as a guarantee.  There still exists a benign
+                    // race condition where a transport producer may be created after the client has been closed; in this case
+                    // the transport producer will be force-closed with the associated connection or, worst case, will close once
+                    // its idle timeout period elapses.
 
-            events = events.ToList();
-            InstrumentMessages(events);
+                    Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));
 
-            try
-            {
-                await activeProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception ex)
-            {
-                scope.Failed(ex);
-                throw;
+                    var pooledItem = PartitionProducerPool.GetPartitionProducer(options.PartitionId, Connection, RetryPolicy);
+
+                    await using var _ = pooledItem.ConfigureAwait(false);
```

Comment:
@jsquire @kinelski I faced the same problem you mentioned [here](https://github.com/Azure/azure-sdk-for-net/pull/9505/files#r367419210) and I came out with this "solution".

Reply:
Yup, the approach seemed reasonable to me;  I wasn't even thinking in terms of synchonization context with the `await using` deal before that PR, but we definitely need to.

Reply:
Neither was I